### DISCUSSION TITLE
Binding to IObservable result from IValueConverter

### DISCUSSION THREAD

**Discussion by josvanouwerkerk** [AvaloniaUI/Avalonia#13794] [Q&A]
Is there any way to have an IValueConverter return an IObservable and then bind to that? The input into the converter is not an IObservable by the way.

So I'm essentially trying to combine:
https://docs.avaloniaui.net/docs/guides/data-binding/how-to-create-a-custom-data-binding-converter
And:
https://docs.avaloniaui.net/docs/guides/data-binding/how-to-bind-to-an-observable

These options don't work:
{Binding Source, Converter={StaticResource Converter}} (Empty result)
{Binding Source^, Converter={StaticResource Converter}} (Doesn't compile)
{Binding Source, Converter={StaticResource Converter}}^ (Empty result)

As to the reason why to do this, I have a StringSource that could be either a direct string or a translation key. So the source is pretty much static, but since the language can change during runtime, the output from the converter is dynamic.

**Comment by timunie** [comment]
better to use a custom markup in that case. https://github.com/AvaloniaUI/Avalonia/discussions/13678

**Comment by josvanouwerkerk** [comment]
I'm not sure this would be the way to go. The source of the translations already is a DynamicResource. What my StringSource attempts to do is enabling you to select between a translation and a static string, mainly for in work progress without having to add everything into the translations. Now both can be represented as IObservable<string>, so that's why I intended to go that route.

I guess I could use a Extension that takes a StringSource as key and then either returns a DynamicResourceExtension or a static value, but that honestly doesn't seem worth the trouble. I might as well use static strings during work in progress and then at some point start treating them as DynamicResource keys and add the translations.

**Reply by timunie** [reply]
You can specify a fallback-dynamicresource in App.xaml or a Resource dict and use that. Add translations later...

**Comment by josvanouwerkerk** [comment]
Well, I decided to give translation a special status in the MVVM model. The intention of the StringSource is to optionally pass a key for translation from the ViewModel. The translation key is technically part of the View, but I don't feel like adding another layer of separation for these keys. So I solved it by making the StringSource (in the ViewModel) an IObservable<string> to which I can just bind in axaml with the stream ^ operator.

```
public abstract record StringSource : IObservable<string>
{
    protected abstract IObservable<string> Source { get; }

    public IDisposable Subscribe(IObserver<string> observer)
    {
        return Source.Subscribe(observer);
    }

    public static implicit operator StringSource(string text)
    {
        return new String(text);
    }

    private record String(string Text) : StringSource
    {
        protected override IObservable<string> Source => Observable.Return(Text);
    }
}
```

Then in the translation namespace there is an override Translate that links to the avalonia resources. So, if the view becomes something completely different, Translate will need an update.

```
public record Translate(string Key) : ViewModelBase.StringSource
{
    protected override IObservable<string> Source
    {
        get
        {
            if (Application.Current is not IResourceHost host)
                return Observable.Return(Key);

            return host.GetResourceObservable<string>(Key);
        }
    }

    public static explicit operator Translate(string key) => new(key);
}

```
So now I can set a StringSource as "Static text" or (Translate)"DynamicResource.Key". Not 100% clean from the MVVM perspective, but I'd say fairly practical and only the Translate record is in the island between the View and the ViewModel.

```
private StringSource header;

header = "Static text";
header = (Translate)"DynamicResource.Key";
```
